### ISSUE TITLE
docker-compose scale causes undefined behavior with Swarm

### ISSUE THREAD

**Issue by nathanleclaire** [docker/compose#2447]
Consider something like:

```yaml
kv:
  image: redis
  environment:
    - affinity:com.docker.compose.service!=kv
```

(I used the Compose-assigned labels for brevity, but the same problem exists if you do it with your own custom labels).

Running a similar command using the `docker` CLI (serially) schedules as expected (will cut you off after all the nodes have containers with the label), but Compose will allow you to `scale` and schedule unexpectedly.  In a 5 node Swarm cluster, no complaints about `docker-compose scale kv=6`:

```console
$ docker-compose scale kv=6
Creating and starting 2 ... done
Creating and starting 3 ... done
Creating and starting 4 ... done
Creating and starting 5 ... done
Creating and starting 6 ... done
```

But FWIW, a subsequent request to `scale` refuses appropriately:

```console
$ docker-compose scale kv=10
Creating and starting 7 ... error
Creating and starting 8 ... error
Creating and starting 9 ... error
Creating and starting 10 ... error

ERROR: for 8  unable to find a node that satisfies com.docker.compose.service!=kv
ERROR: for 9  unable to find a node that satisfies com.docker.compose.service!=kv
ERROR: for 10  unable to find a node that satisfies com.docker.compose.service!=kv
ERROR: for 7  unable to find a node that satisfies com.docker.compose.service!=kv
```

If you bash on it and just chuck `docker-compose scale kv=10` over the wall right from the start, behavior starts to get _really_ erratic:

```
$ docker-compose scale kv=10
Creating and starting 2 ... done
Creating and starting 3 ... done
Creating and starting 4 ... done
Creating and starting 5 ... done
Creating and starting 6 ... error
Creating and starting 7 ... done
Creating and starting 8 ... done
Creating and starting 9 ... done
Creating and starting 10 ... done

ERROR: for 6  Cannot start container 83fe38d5ccd62129527f7ff96604f55760ed3bdf3317164f005f77fe63c4062b: [8] System error: open /sys/fs/cgroup/memory/system.slice/docker-83fe38d5ccd62129527f7ff96604f55760ed3bdf3317164f005f77fe63c4062b.scope/memory.swappiness: no such file or directory
Removing tmp_kv_6 ... done
```

(That last error seems related to Swarm/Docker, not Compose, but it was what I got).

I'm not sure what the best resolution is.  Compose making requests in serial seems inelegant.  Maybe some day Swarm could "stage" / batch scheduled containers for you before actually doing it to mitigate these types of issues.  But at any rate, it would be good to come to a consensus and document what the expected behavior of both tools should be in these scenarios.

**Comment by dnephin** [comment]
I think swarm should error on `docker-compose scale kv=6`. Compose makes an API call per instance, so once it tries to launch instance 6, it should fail, and the error will be reported by compose.

What's the output of `ps` after running scale to 6? Is it actually running 6 containers?

**Comment by nathanleclaire** [comment]
@dnephin Yep.  Note the duplicated schedule on `queenbee`:

```console
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS               NAMES
bbc0977a7279        redis               "/entrypoint.sh redis"   4 seconds ago       Up 3 seconds        6379/tcp            workerbee-2/tmp_kv_4
1a097365c873        redis               "/entrypoint.sh redis"   4 seconds ago       Up 3 seconds        6379/tcp            workerbee-3/tmp_kv_2
599e50fe980f        redis               "/entrypoint.sh redis"   4 seconds ago       Up 2 seconds        6379/tcp            queenbee/tmp_kv_6
50e9e06c9180        redis               "/entrypoint.sh redis"   4 seconds ago       Up 2 seconds        6379/tcp            queenbee/tmp_kv_5
4560354a8143        redis               "/entrypoint.sh redis"   5 seconds ago       Up 3 seconds        6379/tcp            riemann/tmp_kv_3
ce6aef71c705        redis               "/entrypoint.sh redis"   10 seconds ago      Up 8 seconds        6379/tcp            workerbee-1/tmp_kv_1
```

**Comment by dnephin** [comment]
Ok, so that just feels like a bug in swarm. I could just as easily run `parallel docker run ...` and have it create a bunch of containers.  If swarm doesn't return an error when it can't satisfy the constraints, there must be a race condition with the constraint logic?

**Comment by nathanleclaire** [comment]
@dnephin Seems reasonable.  Didn't know if there was something in the Compose integration in particular that might be tripping up the issue.

cc @abronan @vieux @aluzzardi  ^^

**Comment by abronan** [comment]
Problem lies in the scheduler which is no longer sequential on the pull phase. All requests are going to succeed because the labels are not updated effectively until the pull occurs and the containers are created and started effectively.

What happens exactly is that the first batch of scale will not error out because `kv6` has an incomplete view of the labels that are set in the cluster. It does not yet see that there are other containers with a `kv` label set. It's going to fail on the second request though, as now we have a complete view of the labels.

What should be done on the swarm side is create a temporary list of labels applied in the cluster for each request to use until the pulls are completed and the containers are created and started effectively (and remove the labels from the temporary list of labels if the request fails somehow). This gives the occasion to subsequent requests to have a complete view of the labels and take their scheduling decision accordingly (thus on the request to scale and create `kv6`, it will fail).

**Comment by aluzzardi** [comment]
@abronan That's exactly what Swarm is already doing.

Looking into it - there must be an issue on the constraints side.

**Comment by aluzzardi** [comment]
@nathanleclaire @dnephin 

Thanks for the report.

Turns out, it was an error in the parallel scheduling support for Swarm, which was not taking into account labels of containers not yet deployed.

Fixed with docker/swarm#1451 and added an integration test so it won't happen again - feel free to give it a try!

**Comment by nathanleclaire** [comment]
Great news @aluzzardi !
